// rebass/dist/withRebass.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _margin = require('understyle/dist/margin');

var _padding = require('understyle/dist/padding');

var _radii = require('./util/radii');

var _radii2 = _interopRequireDefault(_radii);

var _colorStyle = require('./util/color-style');

var _colorStyle2 = _interopRequireDefault(_colorStyle);

var _config = require('./config');

var _config2 = _interopRequireDefault(_config);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var withRebass = function withRebass(Comp) {
  var RebassBase = function (_React$Component) {
    _inherits(RebassBase, _React$Component);

    function RebassBase() {
      _classCallCheck(this, RebassBase);

      return _possibleConstructorReturn(this, (RebassBase.__proto__ || Object.getPrototypeOf(RebassBase)).apply(this, arguments));
    }

    _createClass(RebassBase, [{
      key: 'render',
      value: function render() {
        var rebass = this.context.rebass;


        var theme = _extends({}, _config2.default, rebass);
        var scale = theme.scale;
        var colors = theme.colors;
        var borderRadius = theme.borderRadius;

        var themeStyle = theme[Comp.name] || {};

        var subComponentStyles = Object.keys(themeStyle).reduce(function (style, key) {
          if (themeStyle[key] && _typeof(themeStyle[key]) === 'object') {
            style[key] = themeStyle[key];
          }
          return style;
        }, {});

        var margin = (0, _margin.setScale)(scale);
        var padding = (0, _padding.setScale)(scale);

        // extract common style props
        var _props = this.props;
        var p = _props.p;
        var pt = _props.pt;
        var pr = _props.pr;
        var pb = _props.pb;
        var pl = _props.pl;
        var px = _props.px;
        var py = _props.py;
        var m = _props.m;
        var mt = _props.mt;
        var mr = _props.mr;
        var mb = _props.mb;
        var ml = _props.ml;
        var mx = _props.mx;
        var my = _props.my;
        var gutter = _props.gutter;
        var rounded = _props.rounded;
        var pill = _props.pill;
        var circle = _props.circle;
        var color = _props.color;
        var backgroundColor = _props.backgroundColor;
        var inverted = _props.inverted;
        var style = _props.style;
        var _props$baseRef = _props.baseRef;
        var baseRef = _props$baseRef === undefined ? function (x) {
          return x;
        } : _props$baseRef;

        var props = _objectWithoutProperties(_props, ['p', 'pt', 'pr', 'pb', 'pl', 'px', 'py', 'm', 'mt', 'mr', 'mb', 'ml', 'mx', 'my', 'gutter', 'rounded', 'pill', 'circle', 'color', 'backgroundColor', 'inverted', 'style', 'baseRef']);

        var sx = _extends({
          boxSizing: 'border-box'
        }, themeStyle, margin({ m: m, mt: mt, mr: mr, mb: mb, ml: ml, mx: mx, my: my, gutter: gutter }, scale), padding({ p: p, pt: pt, pr: pr, pb: pb, pl: pl, px: px, py: py }, scale), (0, _colorStyle2.default)({
          theme: theme,
          color: color,
          backgroundColor: backgroundColor,
          inverted: inverted
        }, colors, theme), (0, _radii2.default)({
          rounded: rounded,
          pill: pill,
          circle: circle
        }, borderRadius), style);

        return _react2.default.createElement(Comp, _extends({}, props, {
          ref: function ref(r) {
            return baseRef(r);
          },
          theme: theme,
          subComponentStyles: subComponentStyles,
          style: sx
        }));
      }
    }]);

    return RebassBase;
  }(_react2.default.Component);

  RebassBase.contextTypes = {
    rebass: _react2.default.PropTypes.object
  };

  return RebassBase;
};

exports.default = withRebass;
